### PR TITLE
Add support for membership inference attacks on seq2seq models

### PR DESCRIPTION
**Pull request by amad-person**
Changes made to the API:
1. Added Seq2SeqAttackInputData that accepts logits, labels for seq2seq datasets.
2. Added create_seq2seq_attacker_data() that generates ranks for the attacker from logits, labels.
3. Added run_seq2seq_attack() to run the average rank thresholding membership inference attack.

A codelab demonstrating the attack on the SATED dataset has been added as well.

### CODE REVIEW

**Comment by amad-person** [comment]
@googlebot I signed it!

File: tensorflow_privacy/privacy/membership_inference_attack/data_structures.py
Comment (dvadym) [inline tensorflow_privacy/privacy/membership_inference_attack/data_structures.py:377]: This function checks that gen is iterator, maybe to name _is_iterator
gen-> obj? (it's not neccasary that the argument is gen, so it's better to name more generic)

File: tensorflow_privacy/privacy/membership_inference_attack/membership_inference_attack_test.py
Comment (dvadym) [inline tensorflow_privacy/privacy/membership_inference_attack/membership_inference_attack_test.py:39]: Nit: get -> Returns

File: tensorflow_privacy/privacy/membership_inference_attack/membership_inference_attack_test.py
Comment (dvadym) [inline tensorflow_privacy/privacy/membership_inference_attack/membership_inference_attack_test.py:43]: Nit: logits_train, labels_train = [], [] (for consistency with line 72)

File: tensorflow_privacy/privacy/membership_inference_attack/models.py
Comment (dvadym) [inline tensorflow_privacy/privacy/membership_inference_attack/models.py:162]: Nit: add type annotation for logits, lables and return argument

File: tensorflow_privacy/privacy/membership_inference_attack/models.py
Comment (dvadym) [inline tensorflow_privacy/privacy/membership_inference_attack/models.py:137]: It looks the only place where train/test_size is needed. But it the same as the number of elements attack_input_train. So we don't need train/test_size.
If that's true, please drop train/test_size from attack_input_data (no need to ask a API user to specify something that it's possible to figure out.

File: tensorflow_privacy/privacy/membership_inference_attack/membership_inference_attack.py
Comment (dvadym) [inline tensorflow_privacy/privacy/membership_inference_attack/membership_inference_attack.py:191]: Cool, that existing code for LogisticRegressionAttacker is reused!

**Review by dvadym — COMMENTED** [review]
Thanks it looks good! I left some comments (minor mostly)
Could you please check that you accepted aggrement on 
https://cla.developers.google.com/ ?

File: tensorflow_privacy/privacy/membership_inference_attack/models.py
Comment (amad-person) [reply tensorflow_privacy/privacy/membership_inference_attack/models.py]: I've used train_size and test_size to print out the attack input shape in the following [part](https://github.com/tensorflow/privacy/pull/137/files#diff-45da7eb7a63fd5e4edabfc2f14124a5e1cde7cffdaf242d424838484de0bff04R439). Since logits and labels are being lazy loaded the user needs to specify train_size and test_size. 

But we could drop train_size and test_size if we don't need to print out the exact attack input shape?

**Comment by CdavM** [comment]
Hi @amad-person!

Just to confirm, have you made all the changes that @dvadym requested? If so, I'll go ahead and merge this this week (@dvadym is out this week). I see there's an open question around the use of train_size - I think we can go ahead and submit as is for now. We can always make more changes later if need be.

**Comment by amad-person** [comment]
Hi @CdavM, I've made all the changes requested (apart from the open question on the train/test_size usage). Thanks!

**Review by CdavM — APPROVED** [review]

**Comment by CdavM** [comment]
Hi @amad-person , one more question: Did you copy any code for the codelab from elsewhere? If so, we need to comply with their licenses (need to check whether this is allowed and whether we need to give them credit somewhere).

**Comment by amad-person** [comment]
I've taken the data loading and model training code from https://github.com/csong27/auditing-text-generation. I added references to the specific files above the respective code cells as well.

**Comment by CdavM** [comment]
Hmm, I'm unfortunately clueless when it comes to these things. I will need to double check with our lawyers to see if adding them to the respective cells is sufficient.

If you'd like, we can merge the rest of the code now, or wait until we hear back for the codelab. Sorry for the delay.

File: tensorflow_privacy/privacy/membership_inference_attack/membership_inference_attack.py
Comment (CdavM) [inline tensorflow_privacy/privacy/membership_inference_attack/membership_inference_attack.py:175]: FYI: The PrivacyReportMetadata is currently unused. Is this intentional?

**Comment by amad-person** [comment]
> Hmm, I'm unfortunately clueless when it comes to these things. I will need to double check with our lawyers to see if adding them to the respective cells is sufficient.
> 
> If you'd like, we can merge the rest of the code now, or wait until we hear back for the codelab. Sorry for the delay.

No problem! We can merge the rest of the code now, and I can open a separate PR with the codelab.

File: tensorflow_privacy/privacy/membership_inference_attack/membership_inference_attack.py
Comment (amad-person) [reply tensorflow_privacy/privacy/membership_inference_attack/membership_inference_attack.py:175]: Oh, I think I missed this. 

I was planning to call `_compute_missing_privacy_report_metadata()` but saw that:
1. it was calculating the accuracy using `logits` and `labels`, but for seq2seq models these are being lazy loaded
2. it was populating `loss_train` and `loss_test`, which are not provided by the user for seq2seq models

I'm actually not sure if I should include PrivacyReportMetadata, any suggestions?

Comment (CdavM) [reply tensorflow_privacy/privacy/membership_inference_attack/membership_inference_attack.py:175]: It sounds like this will be a change significant enough to deserve its own PR to me :-) 

By the way, why are logits and labels being lazily loaded? Is the order of magnitude so large that we can't just store them in memory?

If so, we'll need to change the way we compute the accuracy and loss and perhaps bypass _compute_missing_privacy_report_metadata(). Or have a separate function for seq2seq models?

I can mark the variable unused for now and we can implement the rest in a followup PR. What do you think?

Comment (amad-person) [reply tensorflow_privacy/privacy/membership_inference_attack/membership_inference_attack.py:175]: > By the way, why are logits and labels being lazily loaded? Is the order of magnitude so large that we can't just store them in memory?

Each batch of logits has shape `(num_sequences, num_words, vocab_size)`, so if the `vocab_size` is large enough it might be inefficient to store all batches in memory. Similarly, if each batch has a large number of sequences (and a large number of words in each sequence) storing all the labels at once can become inefficient.

> If so, we'll need to change the way we compute the accuracy and loss and perhaps bypass _compute_missing_privacy_report_metadata(). Or have a separate function for seq2seq models?

Hmm, perhaps we can include a flag for seq2seq models and populate the necessary fields using the existing function, or we can create a separate function `_compute_seq2seq_missing_privacy_report_metadata()` - whichever is cleaner.

> I can mark the variable unused for now and we can implement the rest in a follow-up PR. What do you think?

Sounds good!

**Comment by CdavM** [comment]
Hi @amad-person , a quick update. We heard back from the lawyers and there won't be a problem with adding the codelab, as long as it's in its own directory with the accompanying license file. I'll move it into the appropriate folder.

During an internal review a few more comments popped up. I'm sorry about the hectic process, we're still trying to figure out how to combine the github review process with our internal tools (we need to keep the files on github in sync with the ones at Google). I'm going to copy the comments over now.

File: tensorflow_privacy/privacy/membership_inference_attack/membership_inference_attack.py
Comment (CdavM) [inline tensorflow_privacy/privacy/membership_inference_attack/membership_inference_attack.py:191]: Copying this comment: "Can we add a comment why we don't support LR but not the other attackers?"

File: tensorflow_privacy/privacy/membership_inference_attack/membership_inference_attack.py
Comment (CdavM) [inline tensorflow_privacy/privacy/membership_inference_attack/membership_inference_attack.py:209]: Copying this comment: "Are slices actually supported by this type of attack?

If not, let's add that to the doc and return an error if the user has specified a slice"

File: tensorflow_privacy/privacy/membership_inference_attack/data_structures.py
Comment (CdavM) [inline tensorflow_privacy/privacy/membership_inference_attack/data_structures.py:384]: Copying this comment: "Now we have quite of bit of seq2sec-specific logic in files that were supposed to be generic (data_structures, models, membership_inference_attack), which makes those files quite heavy and much less generic than we hoped they'd be.

I wonder if this makes sense to put all seq2seq specific logic in a separate file (seq2seq.py), which will include pieces from the 3 files above.

WDYT?

This is food for thought, not blocking this PR on this discussion which might require a broader revamp of our file structure."

File: tensorflow_privacy/privacy/membership_inference_attack/membership_inference_attack.py
Comment (amad-person) [reply tensorflow_privacy/privacy/membership_inference_attack/membership_inference_attack.py:191]: Currently the seq2seq attack has a single number (average rank) in each record of the attack datasets, so I've used the simplest model i.e. LR to find the optimal threshold. 

Should we add support for the other attackers (MLP, RF, KNN) for single number features?

File: tensorflow_privacy/privacy/membership_inference_attack/membership_inference_attack.py
Comment (amad-person) [reply tensorflow_privacy/privacy/membership_inference_attack/membership_inference_attack.py:209]: No, slices aren't supported by this type of attack. The default value of SlicingSpec was "all" so I included it in the attack result 😅. 

> If not, let's add that to the doc and return an error if the user has specified a slice.

Sure, I will update the PR with the specified changes.

File: tensorflow_privacy/privacy/membership_inference_attack/data_structures.py
Comment (amad-person) [reply tensorflow_privacy/privacy/membership_inference_attack/data_structures.py:384]: Hmm, it would be a good idea to separate out the logic for seq2seq models since we could potentially add more attack types (like training the attacker on histograms of ranks).

File: tensorflow_privacy/privacy/membership_inference_attack/membership_inference_attack.py
Comment (sushkoy) [reply tensorflow_privacy/privacy/membership_inference_attack/membership_inference_attack.py:191]: My 50c: I think it's OK to keep linear regression, since it makes most sense for the single number features. I'd vote for dropping a comment explaining that for posterity, e.g.:

# Only linear regressions are supported, since they make most sense for single-number features

**Comment by CdavM** [comment]
Thanks for making the changes, @amad-person ! Can you let us know when you're done with the rest of the changes and the PR is ready?

**Comment by amad-person** [comment]
@CdavM wasn't sure if I should I create `seq2seq.py` or is that going to be done in a future PR?

**Comment by CdavM** [comment]
That can be done in a future PR. But I think slicing still has to be
addressed?

**Comment by amad-person** [comment]
Thanks! Yup, I'll push the slicingspec changes and let you know.

**Comment by amad-person** [comment]
Hi @CdavM, I updated the PR with the slicingspec changes, but I think I'll need some clarifications before we can merge the PR.

Change: I removed the call to `_get_slicing_spec()`.

The `run_seq2seq_attack()` function returns a `SingleAttackResult` object, and I realised that `SingleAttackResult` expects a `SingleSliceSpec` attribute. So I had to pass in the default `SingleSliceSpec` (which corresponds to no slice on the dataset, all records are returned).

Also, the `run_seq2seq_attack()` function doesn't accept a `SlicingSpec` parameter so the user can't pass in any slicing specification. Is there a need to return an error message in this case?

**Comment by CdavM** [comment]
Hi @amad-person !

If we don't do any slicing, it's appropriate to return the SliceSpec for the entire dataset. 

I don't think we need to return an error if we don't use the SlicingSpec parameter for now. Does slicing make sense for seq2seq models? If so, we can extend it later. I think we'll need a follow-up PR in any case.

**Comment by amad-person** [comment]
@CdavM 

> If we don't do any slicing, it's appropriate to return the SliceSpec for the entire dataset.

Thanks! I think I've addressed the required changes for this PR then.

> Does slicing make sense for seq2seq models? 

The attack I included (average rank thresholding) doesn't use slicing. I'm not sure if we can use slicing for attacks on seq2seq models. Perhaps slicing = number of sentences included in a batch to calculate the average rank? 

> If so, we can extend it later. I think we'll need a follow-up PR in any case.

Sure!

**Review by CdavM — APPROVED** [review]

**Comment by CdavM** [comment]
Okay, we finally merged this PR :-) Thanks again for your patience!

The next step is to do some refactoring and move all the seq2seq specific logic into a separate file. Any other things that need to be done?

It sounds like we should discuss a natural way to slice the data in a seq2seq model before implementing that.

**Comment by amad-person** [comment]
Thanks @CdavM! 🎉 

From our discussion on this PR, we have the following things to do:
1. Populating PrivacyReportMetadata for seq2seq models ([comment thread](https://github.com/tensorflow/privacy/pull/137#discussion_r520671276)).
2. Refactoring seq2seq logic and tests into separate files.
3. Defining and then supporting slicing for seq2seq data.

Perhaps we could schedule a meeting to discuss these updates?

**Comment by CdavM** [comment]
Hi @amad-person,

Sounds good - I sent you a meeting invite for the following Monday.

**Comment by amad-person** [comment]
Thanks @CdavM! Will join.

**Comment by shs037** [comment]
Hi @amad-person , I have one question regarding [seq2seq_mia.py](https://github.com/tensorflow/privacy/blob/master/tensorflow_privacy/privacy/privacy_tests/membership_inference_attack/seq2seq_mia.py#L351): since the feature is one scalar value, why don't we use the [threshold attack](https://github.com/tensorflow/privacy/blob/master/tensorflow_privacy/privacy/privacy_tests/membership_inference_attack/membership_inference_attack.py#L94)? I think logistic regression would work in the same way as thresholding here as it's a linear classifier. Am I missing something?

Thanks in advance for any help!

**Comment by amad-person** [comment]
Hi @shs037, yes you are right about the logistic regression model working in the same way as the threshold attack. I used the logistic regression model solely because of implementation choices. 

Using the [threshold attack](https://github.com/tensorflow/privacy/blob/master/tensorflow_privacy/privacy/privacy_tests/membership_inference_attack/membership_inference_attack.py#L94) function requires the object [AttackInputData](https://github.com/tensorflow/privacy/blob/3a4c4400a682e2dc606987f53f6f544244fb8c7a/tensorflow_privacy/privacy/privacy_tests/membership_inference_attack/data_structures.py#L170) to be passed as an argument, and it calls the `.get_loss_train()` and `.get_loss_test()` methods of this object to get the required loss values for the attack. 

However for the seq2seq attack, I can't use the same method because we need to compute the ranks from the logits and labels instead of loss values. So I created a new data object [Seq2SeqAttackInputData](https://github.com/tensorflow/privacy/blob/3a4c4400a682e2dc606987f53f6f544244fb8c7a/tensorflow_privacy/privacy/privacy_tests/membership_inference_attack/seq2seq_mia.py#L50) which generates and stores this information, and created an analogous function [run_seq2seq_attack](https://github.com/tensorflow/privacy/blob/3a4c4400a682e2dc606987f53f6f544244fb8c7a/tensorflow_privacy/privacy/privacy_tests/membership_inference_attack/seq2seq_mia.py#L330) to run the threshold attack.

Please let me know if you have any further questions!

**Comment by shs037** [comment]
Thanks for your reply!
Is it possible to do something like:
1. computing the ranks in `rank_train` and `rank_test`
2. call the threshold attack with loss set to the rank:
```
input = AttackInputData(
  loss_train = rank_train,
  loss_test = rank_test,
)
attacks_result = mia.run_attacks(input, attack_types = [AttackType.THRESHOLD_ATTACK])
```

**Comment by amad-person** [comment]
@shs037 yes I think this should work!

**Comment by shs037** [comment]
Hi @amad-person, I updated the code in https://github.com/tensorflow/privacy/commit/ceced43d0b158758cedcecbbad8a6a8f997883ec. Please feel free to take a look and let me know if it makes sense. And of course also feel free to make any changes.

**Pull request merged**
